pkg/tls/config.go

```diff
@@ -0,0 +1,214 @@
+	if err := cfg.update(); err != nil {
+		return nil, err
+	}
+	if err := cfg.start(); err != nil {
+		return nil, err
+	}
+
+	return cfg, nil
+}
+
+// start initialize the WatchedConfig files watcher and update goroutine.
+func (cfg *WatchedConfig) start() error {
+	// file watcher setup.
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		return fmt.Errorf("Failed to create fsnotify watcher: %s", err)
+	}
+	for _, path := range cfg.caFiles {
+		if err := watcher.Add(path); err != nil {
+			watcher.Close()
+			return fmt.Errorf("Failed to add %q to fsnotify watcher: %s", path, err)
+		}
+	}
+	if cfg.certFile != "" {
+		if err := watcher.Add(cfg.certFile); err != nil {
+			watcher.Close()
+			return fmt.Errorf("Failed to add %q to fsnotify watcher: %s", cfg.certFile, err)
+		}
+	}
+	if cfg.privkeyFile != "" {
+		if err := watcher.Add(cfg.privkeyFile); err != nil {
+			watcher.Close()
+			return fmt.Errorf("Failed to add %q to fsnotify watcher: %s", cfg.privkeyFile, err)
+		}
+	}
+
+	go func() {
+		defer watcher.Close()
+		for {
+			select {
+			case event := <-watcher.Events:
+				cfg.log.Debugf("Received fsnotify event: %+v", event)
+				switch event.Op {
+				case fsnotify.Create, fsnotify.Write, fsnotify.Chmod, fsnotify.Remove, fsnotify.Rename:
+					if err := cfg.update(); err != nil {
+						cfg.log.WithError(err).Warn("config update failed")
+					}
+				default:
+					cfg.log.Warnf("Watcher received unknown event: %s. Ignoring.", event)
+				}
+			case err := <-watcher.Errors:
+				cfg.log.WithError(err).Warn("Watcher received an error")
+			case <-cfg.stop:
+				cfg.log.Info("Stopping")
+				return
+			}
+		}
+	}()
+
+	return nil
+}
+
+// CertificateAuthorityConfigured returns true when the WatchedConfig has at
+// least one CA file, false otherwise.
+func (cfg *WatchedConfig) CertificateAuthorityConfigured() bool {
+	return len(cfg.caFiles) > 0
+}
+
+// KeypairConfigured returns true when the WatchedConfig contains both a
+// certificate and its private key, false otherwise.
+func (cfg *WatchedConfig) KeypairConfigured() bool {
+	return cfg.certFile != "" && cfg.privkeyFile != ""
+}
+
+// update read the WatchedConfig files.
+func (cfg *WatchedConfig) update() error {
+	cfg.mu.Lock()
+	defer cfg.mu.Unlock()
+
+	// CA update
+	if cfg.CertificateAuthorityConfigured() {
+		if len(cfg.caFiles) > 0 {
+			caCertPool := x509.NewCertPool()
+			for _, path := range cfg.caFiles {
+				pem, err := ioutil.ReadFile(path)
+				if err != nil {
+					return fmt.Errorf("failed to load cert %q: %s", path, err)
+				}
+				if ok := caCertPool.AppendCertsFromPEM(pem); !ok {
+					return fmt.Errorf("failed to load cert %q: must be PEM encoded", path)
+				}
+			}
+			cfg.caCertPool = caCertPool
+		} else {
+			caCertPool, err := x509.SystemCertPool()
+			if err != nil {
+				return fmt.Errorf("failed to load the system ca certificates: %s", err)
+			}
+			cfg.caCertPool = caCertPool
+		}
+	}
+
+	// keypair update
+	if cfg.KeypairConfigured() {
```

Comment:
Related to the above comment, as all files are reloaded with the `update()` method, if re-loading the CAs fails, then the keypair will not be reloaded. This is a bit unfortunate if it turns out that the`fsnotify.Event` concerned one of the `certFile` or `privkeyFile` but not any of the `caFiles`.

Reply:
If we successfully loaded the CA and keypair previously, and receive an event about the keypair, then (re)loading the CA should work right? Granted you could get some I/O issue etc. but that's out of control anyway.